presto-hudi/src/main/java/com/facebook/presto/hudi/RecordCursorUtil.java

```diff
@@ -0,0 +1,162 @@
+                split.getTable().getSchemaName(),
+                split.getTable().getTableName(),
+                split.getPath(),
+                false);
+        Configuration configuration = hdfsEnvironment.getConfiguration(context, path);
+        return hdfsEnvironment.doAs(session.getUser(), () -> {
+            RecordReader<?, ?> recordReader = createRecordReader(configuration, schema, split, dataColumns);
+            @SuppressWarnings("unchecked") RecordReader<?, ? extends Writable> reader = (RecordReader<?, ? extends Writable>) recordReader;
+            return createRecordCursor(configuration, path, reader, split.getLength(), schema, dataColumns, hiveStorageTimeZone, typeManager);
+        });
+    }
+
+    private static RecordReader<?, ?> createRecordReader(
+            Configuration configuration,
+            Properties schema,
+            HudiSplit split,
+            List<HudiColumnHandle> dataColumns)
+    {
+        // update configuration
+        configuration.setBoolean(READ_ALL_COLUMNS, false);
+        configuration.set(READ_COLUMN_IDS_CONF_STR, join(dataColumns, HudiColumnHandle::getId));
+        configuration.set(READ_COLUMN_NAMES_CONF_STR, join(dataColumns, HudiColumnHandle::getName));
+        schema.stringPropertyNames()
+                .forEach(name -> configuration.set(name, schema.getProperty(name)));
+        refineCompressionCodecs(configuration);
+
+        // create input format
+        String inputFormatName = split.getPartition().getStorage().getStorageFormat().getInputFormat();
+        InputFormat<?, ?> inputFormat = createInputFormat(configuration, inputFormatName);
+
+        // create record reader for split
+        try {
+            JobConf jobConf = new JobConf(configuration);
+            Path path = new Path(split.getPath());
+            FileSplit fileSplit = new FileSplit(path, split.getStart(), split.getLength(), (String[]) null);
+            List<HoodieLogFile> logFiles = split.getLogFiles().stream().map(HoodieLogFile::new).collect(toList());
+            String tablePath = split.getTable().getPath();
+            FileSplit hudiSplit = new HoodieRealtimeFileSplit(fileSplit, tablePath, logFiles, split.getInstantTime(), Option.empty());
+            return inputFormat.getRecordReader(hudiSplit, jobConf, Reporter.NULL);
+        }
+        catch (IOException e) {
+            String msg = format("Error opening Hive split %s (offset=%s, length=%s) using %s: %s",
+                    split.getPath(),
+                    split.getStart(),
+                    split.getLength(),
+                    inputFormatName,
+                    firstNonNull(e.getMessage(), e.getClass().getName()));
+            throw new PrestoException(HUDI_CANNOT_OPEN_SPLIT, msg, e);
+        }
+    }
+
+    private static InputFormat<?, ?> createInputFormat(Configuration conf, String inputFormat)
+    {
+        try {
+            Class<?> clazz = conf.getClassByName(inputFormat);
```

Comment:
Instead of reflection-based check, we can use annotations provided by Hudi. Check these two annotations in Hudi:
```
UseRecordReaderFromInputFormat
UseFileSplitsFromInputFormat
```

Reply:
The annotations `UseRecordReaderFromInputFormat` and `UseFileSplitsFromInputFormat` indicate that we could delegate the splitting to InputFormat. As we discussed in the design proposal, there are some performance issues in the splitting implementation from InputFormat. So we will ignore these annotations and provide a new splitting implementation optimized for Presto.